### PR TITLE
Writing SNMP Apps in Python

### PR DESCRIPTION
**Pull request by snmplaboratories**
Please, pull my PyCon paper.

### CODE REVIEW

File: 2015/presentations/writing-snmp-apps/text.md [hunk 1]
```diff
@@ -0,0 +1,470 @@
...
+Network management is important for keeping your ever growing network
+infrastructure healthy and secure. SNMP is a well established
+technology designed to identify and access key operational metrics of
+networked hosts, routers and all sorts of industrial equipment.
+
+As we all know, many organizations around the world use Python for
+their IT automation needs, including network management. Over the
+course of last decade, many SNMP implementations appeared. Some are
+Python bindings to C-based [Net-SNMP library](http://www.net-snmp.org),
+which is by many considered being a reference implementation for SNMP
```
Comment (soltysh) [inline 2015/presentations/writing-snmp-apps/text.md]: which is considered by many a reference...

File: 2015/presentations/writing-snmp-apps/text.md [hunk 2]
```diff
@@ -0,0 +1,470 @@
...
+networked hosts, routers and all sorts of industrial equipment.
+
+As we all know, many organizations around the world use Python for
+their IT automation needs, including network management. Over the
+course of last decade, many SNMP implementations appeared. Some are
+Python bindings to C-based [Net-SNMP library](http://www.net-snmp.org),
+which is by many considered being a reference implementation for SNMP 
+technology. Others are pure-Python modules addressing specific SNMP features.
+
+Among many SNMP libraries in existence in the Python landscape, right
```
Comment (soltysh) [inline 2015/presentations/writing-snmp-apps/text.md]: in existence/existing

File: 2015/presentations/writing-snmp-apps/text.md [hunk 3]
```diff
@@ -0,0 +1,470 @@
...
+from the start, PySNMP project aims at complete and universal SNMP
+implementation offering its users full power of SNMP technology across
```
Comment (soltysh) [inline 2015/presentations/writing-snmp-apps/text.md]: comma after implementation

File: 2015/presentations/writing-snmp-apps/text.md [hunk 4]
```diff
@@ -0,0 +1,470 @@
...
+    elif errorStatus:
+        print('%s at %s' % (errorStatus.prettyPrint(),
+                            varBinds[int(errorIndex)-1][0]
+                            if errorIndex else '?'))
+    else:
+        for varBind in varBinds:
+            print(' = '.join([x.prettyPrint() for x in varBind ]))
+
+That code is somewhat verbose for a reason: PySNMP API exposes many SNMP
+details to programmer giving her great power and flexibility (attentive
```
Comment (soltysh) [inline 2015/presentations/writing-snmp-apps/text.md]: exposes many SNMP details, giving programmer great power and flexibility

File: 2015/presentations/writing-snmp-apps/text.md [hunk 5]
```diff
@@ -0,0 +1,470 @@
...
+readers may have spotted a Python generator in the code). But before we
+dive into the details let me remind our readers basic SNMP design and how
+PySNMP architecture maps into it.
+
+A bit of background
+-------------------
+
+Back in early days of computer networking, as local networks grew in size
```
Comment (soltysh) [inline 2015/presentations/writing-snmp-apps/text.md]: in the early

File: 2015/presentations/writing-snmp-apps/text.md [hunk 6]
```diff
@@ -0,0 +1,470 @@
...
+and complexity, keeping an eye on expanding farm of computers, applications
+and other network equipment became a hassle to system administrators.
+Besides simple ping-like methods of testing hosts and services
```
Comment (soltysh) [inline 2015/presentations/writing-snmp-apps/text.md]: methods for testing hosts..

File: 2015/presentations/writing-snmp-apps/text.md [hunk 7]
```diff
@@ -0,0 +1,470 @@
...
+At that time there was no single predominant packet switching technology
+like we see today. Large organizations were busy developing their own
+network stacks each equipped with some form of network management
+facilities. Ultimately, in early 90's, Internet Protocol Suite wins this
+competition practically eliminating all other rivals. For TCP/IP network
+management they offered SNMP, which remains principal technology even
+today.
+
+As a technology, SNMP defines application layer protocol, data model and
+data objects. The protocol, which evolved a great deal since its initial
```
Comment (soltysh) [inline 2015/presentations/writing-snmp-apps/text.md]: which evolved greatly since its...

File: 2015/presentations/writing-snmp-apps/text.md [hunk 8]
```diff
@@ -0,0 +1,470 @@
...
+variables from hosts or applications being managed. 
+
+With SNMP architecture, managed host or application should have a component
+called SNMP Agent. It acts as an intermediate having access to
+host/application internals and being able to funnel that information over
+SNMP to interested parties in form of SNMP variables. 
+
+The other part of the system is called SNMP Manager, this component is
+always looking for SNMP variables either by querying SNMP Agents or
+listening for notifications they may produce whenever something happens to
```
Comment (soltysh) [inline 2015/presentations/writing-snmp-apps/text.md]: happens ~~to them~~

File: 2015/presentations/writing-snmp-apps/text.md [hunk 9]
```diff
@@ -0,0 +1,470 @@
...
+SNMP-level data processing is performed by a collection of SNMP Message 
+Processing ([RFC3412](http://www.ietf.org/rfc/rfc3412.txt)) and Security
+([RFC3414](http://www.ietf.org/rfc/rfc3414.txt)) modules living in 
+*pysnmp.proto...* sub-package. All crypto operations are offloaded to
+the third-party [PyCrypto](https://www.dlitz.net/software/pycrypto/)
+package.
+
+Base classes acting as a wireframe for SMI objects
+([RFC2587](http://www.ietf.org/rfc/rfc2578.txt)) are defined in
+*pysnmp.smi.*... They carry out both of MIB purposes:
```
Comment (soltysh) [inline 2015/presentations/writing-snmp-apps/text.md]: I don't fully understand the sentence starting above, as if something was missing there.

File: 2015/presentations/writing-snmp-apps/text.md [hunk 10]
```diff
@@ -0,0 +1,470 @@
...
+Library orientation
+-------------------
+
+The PySNMP library is internally structured along the lines of 
+[RFC3411](http://www.ietf.org/rfc/rfc3411.txt). Components that are not
+unique to SNMP are put into stand-alone Python packages to promote
+reusability.
+
+SNMP protocol is defined in terms of ASN.1 data structures, SNMP messages
+travelling the wire are encoded in BER. For those purposes PySNMP relies on
```
Comment (soltysh) [inline 2015/presentations/writing-snmp-apps/text.md]: Can you link to rfc/doc describing BER?

File: 2015/presentations/writing-snmp-apps/text.md [hunk 11]
```diff
@@ -0,0 +1,470 @@
...
+and response processed. Generator functions are specific to SNMP message
+type and are uniformly initialized with:
+
+* SNMP Engine object: this is the umbrella object coordinating all
+  SNMP operations. It's used by SNMP applications like Command Generator
+  application featured in example.
+* SNMP authentication method: that can be SNMPv1/v2c Community Name or
+  [RFC3414](http://www.ietf.org/rfc/rfc3414.txt) *UsmUser* object
+  conveying USM username, encryption and ciphering keys.
+* Kind of I/O to use for this communication, endpoints addresses and
```
Comment (soltysh) [inline 2015/presentations/writing-snmp-apps/text.md]: Kind/Type

Comment (soltysh) [inline 2015/presentations/writing-snmp-apps/text.md]: for ~~this~~ communication

File: 2015/presentations/writing-snmp-apps/text.md [hunk 12]
```diff
@@ -0,0 +1,470 @@
...
+    ...
+
+An efficient Manager could then optimize its behavior by fetching required
+MIBs and querying MIB variables from those MIBs.
+
+SNMP Agents
+-----------
+
+PySNMP could act as a full-blown SNMP Agent supporting most of standard
+SNMP features pretty much out out the box. When building SNMP Agent, most
```
Comment (soltysh) [inline 2015/presentations/writing-snmp-apps/text.md]: out ~~out~~ of the box

**Comment by soltysh** [comment]
@snmplaboratories thanks for the article, mostly it looks good, but I have one doubt I need you to explain. Other than that please squash your commits into one commit.

**Comment by snmplaboratories** [comment]
Thanks and sorry for that many errors! I will apply your fixes shortly 
and squash commits as you requested.

Please, let me know what needs an explanation from my side.

**Comment by snmplaboratories** [comment]
hopefully, done.

**Comment by soltysh** [comment]
LGTM.

**Comment by soltysh** [comment]
Thanks Ilya!

**Comment by snmplaboratories** [comment]
Hi Maciej,

I've done some minor fixes to the paper in my repo. If it is not too 
late, it would make sense to merge.

Thanks!

**Pull request merged**